src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -0,0 +1,351 @@
+                            return CheckStatement(statement);
+
+                        default:
+                            // If the current node was neither an expression nor a statement,
+                            // it's probably a part of another statement like a local declaration
+                            // initializer, so we just look up for a parenting statement.
+                            return CheckStatement(current.GetAncestor<StatementSyntax>());
+                    }
+                }
+
+                return false;
+            }
+
+            private bool CheckLoop(SyntaxNode statement, StatementSyntax body, bool defAssignedWhenTrue)
+            {
+                if (_operand.Kind() == SyntaxKind.IdentifierName)
+                {
+                    // We have something like:
+                    //
+                    //      var x = e as T;
+                    //      while (b != null) { ... }
+                    //
+                    // It's not necessarily safe to convert this to:
+                    //
+                    //      while (x is T b) { ... }
+                    //
+                    // That's because in this case, unlike the original code, we're
+                    // type-checking in every iteration, so we do not replace a
+                    // simple null check with the "is" operator if it's in a loop.
+                    return false;
+                }
+
+                if (!defAssignedWhenTrue && LocalFlowsIn(body))
+                {
+                    // If the local is accessed before assignment
+                    // in the loop body, we should make sure that
+                    // the variable is definitely assigned by then.
+                    return false;
+                }
+
+                // The scope of the pattern variables for loops
+                // does not leak out of the loop statement.
+                return !IsAccessedOutOfScope(scope: statement);
+            }
+
+            private bool CheckExpression(ExpressionSyntax exprsesion)
+            {
+                // It wouldn't be safe to read after the pattern variable is
+                // declared inside a sub-expression, because it would not be
+                // definitely assigned after this point. It's possible to allow
+                // use after assignment but it's rather unlikely to happen.
+                return !IsAccessedOutOfScope(scope: exprsesion);
+            }
+
+            private bool CheckStatement(StatementSyntax statement)
+            {
+                Debug.Assert(statement != null);
+
+                // This is either an embedded statement or parented by a block.
+                // If we're parented by a block, then that block will be the scope
+                // of the new variable. Otherwise the scope is the statement itself.
+                if (statement.Parent.IsKind(SyntaxKind.Block, out BlockSyntax block))
```

Comment:
is this always true?  what about embedded statements?   (i honestly don't know.  but i just want to make sure we thoughta bout it).

Reply:
There's an `else` that cover embedded statements.